README.md
# flutter_logcat_monitor

Flutter plugin to monitor the stream of system messages, stack traces etc using *logcat* command-line tool.

NOTE: This plugin fetches logcat logs only on Android Devices.

# how to use

1. Create a function to consume the *logcat* messages
```dart
  void _mylistenStream(dynamic value) {
    if (value is String) {
      _logBuffer.writeln(value);
    }
  }
```

2. Register your function as a listener to get logs then use it in anyway within your app.
```dart
    FlutterLogcatMonitor.addListen(_mylistenStream);
```

3. Start the logcat monitor passing the [filter parameters](#logcat_filter) as defined in *logcat* tool.
```dart
    await FlutterLogcatMonitor.startMonitor("*.*");
```

## Installation

Need permission in AndroidManifest.xml:
```xml
<uses-permission android:name="android.permission.READ_LOGS" />
```

[https://pub.dev/packages/flutter_logcat_monitor/install](https://pub.dev/packages/flutter_logcat_monitor/install)

# example

See the example code in [example](https://github.com/josephcrowell/flutter_logcat_monitor/tree/main/example) folder.

Here we use a StringBuffer to store the messages and display them on a log screen.

<p align='center'>
    <img src="https://raw.githubusercontent.com/josephcrowell/flutter_logcat_monitor/main/doc/example1.jpg" />
</p>

# under the hood

The FlutterLogcatMonitorPlugin runs event/method channel handlers in **UI-thread** and the *logcat* process monitor in a **background thread** as recomended by Google to not block the UI interface.

<p align='center'>
    <img src="https://raw.githubusercontent.com/josephcrowell/flutter_logcat_monitor/main/doc/diagram.png" />
</p>


### logcat filter options

<a name="logcat_filter"></a>

from the [Android Developers logcat documentation](https://developer.android.com/tools/logcat#filteringOutput):
```txt
filterspecs are a series of
  <tag>[:priority]

where <tag> is a log component tag (or * for all) and priority is:
  V    Verbose (default for <tag>)
  D    Debug (default for '*')
  I    Info
  W    Warn
  E    Error
  F    Fatal
  S    Silent (suppress all output)

'*' by itself means '*:D' and <tag> by itself means <tag>:V.
If no '*' filterspec or -s on command line, all filter defaults to '*:V'.
eg: '*:S <tag>' prints only <tag>, '<tag>:S' suppresses all <tag> log messages.
```

Examples:
 - `*.*` show ALL tags and priorities.
 - `flutter,FlutterLogcatMonitorPlugin,S:*` show flutter and FlutterLogcatMonitorPlugin and suppresses all others.


CHANGELOG.md
# Changelog

## 1.0.2

* Update license

## 1.0.1

* Package and build system updates

## 1.0.0

* Convert to kotlin and a modern threading API
* Code cleanup
* Fork and change namespace
* Update example

## 0.0.6

* Update documentation

## 0.0.5

* Update to null-safety

## 0.0.4

* Update documentation

## 0.0.3

* Update documentation

## 0.0.2

* Update documentation
* Cleaned the code

## 0.0.1

* First working release


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_logcat_monitor_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && (widget.data?.startsWith('Running on:') ?? false),
      ),
      findsOneWidget,
    );
  });
}


test/logcat_monitor_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:logcat_monitor/logcat_monitor.dart';

void main() {
  const MethodChannel channel = MethodChannel('logcat_monitor/methods');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(channel, (MethodCall methodCall) async {
      switch (methodCall.method) {
        case 'startMonitor':
          return true;
        case 'stopMonitor':
          return true;
        case 'runLogcat':
          return 'logcat example';
        default:
          return 'not implemented';
      }
    });
  });

  tearDown(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(channel, null);
  });

  test('getLogcatDump', () async {
    expect(await LogcatMonitor.getLogcatDump, 'logcat example');
  });
}


test/flutter_logcat_monitor_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_logcat_monitor/flutter_logcat_monitor.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_logcat_monitor/methods');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(channel, (MethodCall methodCall) async {
      switch (methodCall.method) {
        case 'startMonitor':
          return true;
        case 'stopMonitor':
          return true;
        case 'runLogcat':
          return 'logcat example';
        default:
          return 'not implemented';
      }
    });
  });

  tearDown(() {
    TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger
        .setMockMethodCallHandler(channel, null);
  });

  test('getLogcatDump', () async {
    expect(await FlutterLogcatMonitor.getLogcatDump, 'logcat example');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_logcat_monitor/flutter_logcat_monitor.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  StringBuffer _logBuffer = StringBuffer();
  int _groupValue = 0;

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  Future<void> initPlatformState() async {
    try {
      debugPrint('Attempting to add listen Stream of log.');
      FlutterLogcatMonitor.addListen(_listenStream);
    } on PlatformException {
      debugPrint('Failed to listen Stream of log.');
    }
    await FlutterLogcatMonitor.startMonitor("*.*");
  }

  void _listenStream(dynamic value) {
    if (value is String) {
      if (mounted) {
        setState(() {
          _logBuffer.writeln(value);
        });
      } else {
        _logBuffer.writeln(value);
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Flutter Logcat Monitor Example App'),
        ),
        body: Column(
          children: [
            Padding(
              padding: EdgeInsets.all(4),
              child: Text("Logcat log:"),
            ),
            logBoxBuild(context),
            RadioListTile(
              title: Text("logcat filter: *.*"),
              value: 0,
              groupValue: _groupValue,
              onChanged: (int? value) async {
                setState(() {
                  _groupValue = value!;
                  _logBuffer.clear();
                });
                await FlutterLogcatMonitor.startMonitor("*.*");
              },
            ),
            RadioListTile(
              title:
                  Text("logcat filter: flutter,FlutterLogcatMonitorPlugin,S:*"),
              value: 1,
              groupValue: _groupValue,
              onChanged: (int? value) async {
                setState(() {
                  _groupValue = value!;
                  _logBuffer.clear();
                });
                await FlutterLogcatMonitor.startMonitor(
                    "flutter:*,FlutterLogcatMonitorPlugin:*,*:S");
              },
            ),
            TextButton(
              child: Text("call debugPrint on flutter"),
              onPressed: () async {
                debugPrint("called debugPrint from flutter!");
              },
              style: TextButton.styleFrom(
                  elevation: 2, backgroundColor: Colors.amber[100]),
            ),
            TextButton(
              child: Text("Clear"),
              onPressed: () async {
                clearLog();
              },
              style: TextButton.styleFrom(
                  elevation: 2, backgroundColor: Colors.amber[100]),
            ),
          ],
        ),
      ),
    );
  }

  Widget logBoxBuild(BuildContext context) {
    return Expanded(
      child: Container(
        alignment: AlignmentDirectional.topStart,
        decoration: BoxDecoration(
          color: Colors.black,
          border: Border.all(
            color: Colors.blueAccent,
            width: 1.0,
          ),
        ),
        child: SingleChildScrollView(
          reverse: true,
          scrollDirection: Axis.vertical,
          child: Text(
            _logBuffer.toString(),
            textDirection: TextDirection.ltr,
            textAlign: TextAlign.start,
            style: TextStyle(
              color: Colors.white,
              fontSize: 14.0,
            ),
          ),
        ),
      ),
    );
  }

  void clearLog() async {
    FlutterLogcatMonitor.clearLogcat;
    await Future.delayed(const Duration(milliseconds: 100));
    setState(() {
      _logBuffer.clear();
      _logBuffer.writeln("log buffer cleared");
    });
  }
}


lib/flutter_logcat_monitor.dart
import 'dart:async';

import 'package:flutter/services.dart';

class FlutterLogcatMonitor {
  static const MethodChannel _channel =
      const MethodChannel('flutter_logcat_monitor/methods');
  static const EventChannel _stream =
      const EventChannel('flutter_logcat_monitor/events');
  static late StreamSubscription _streamSubscription;

  /// Adds a subscription to this stream.
  static void addListen(void Function(dynamic) onData) {
    _streamSubscription = _stream.receiveBroadcastStream().listen(onData);
  }

  /// Cancels a subscription to this stream.
  static void cancelListen() {
    _streamSubscription.cancel();
  }

  /// Runs a logcat command and starts monitoring it through the event stream.
  static Future<bool?> startMonitor(String options) async {
    bool? result = await _channel
        .invokeMethod('startMonitor', <String, String>{'options': options});
    return result;
  }

  /// Interrupt the logcat command and stop monitoring it through the event stream.
  static Future<bool?> stopMonitor() async {
    bool? result = await _channel.invokeMethod('stopMonitor');
    return result;
  }

  /// Runs a logcat command with options and returns the result.
  ///
  /// use -d option to "Dump the log and then exit (don't block)."
  static Future<String?> runLogcat(String options) async {
    final String? result = await _channel
        .invokeMethod('runLogcat', <String, String>{'options': options});
    return result;
  }

  /// Runs a logcat log dump command and returns the result.
  static Future<String?> get getLogcatDump async {
    final String? result = await _channel
        .invokeMethod('runLogcat', <String, String>{'options': "-d"});
    return result;
  }

  /// Runs a logcat clear command and returns the result.
  static Future<String?> get clearLogcat async {
    final String? result = await _channel
        .invokeMethod('runLogcat', <String, String>{'options': "-c"});
    return result;
  }
}


android/src/main/kotlin/com/josephcrowell/logcat_monitor/LogcatMonitorPlugin.kt
package com.josephcrowell.logcat_monitor

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class LogcatMonitorPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private val uiThreadHandler = Handler(Looper.getMainLooper())
    private var channel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventSink? = null
    private var logcatProcess: Process? = null

    companion object {
        private const val TAG_NAME = "LogcatMonPlugin"
        private const val sleepIntervalThread: Long = 1000
        private const val sleepTimeThread: Long = 200
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "logcat_monitor/methods")
        channel!!.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "logcat_monitor/events")
        eventChannel!!.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        var options: String?
        val log = StringBuffer()
        options = call.argument("options")
        if (options.isNullOrEmpty()) {
            options = ""
        }
        when (call.method) {
            "startMonitor" -> {
                closePrevious()
                Log.d(TAG_NAME, "run startMonitor()")
                Thread {
                    logcatMonitor(options)
                }.start()
                result.success(true)
            }

            "stopMonitor" -> {
                Log.d(TAG_NAME, "run stopMonitor()")
                closePrevious()
                result.success(true)
            }

            "runLogcat" -> {
                Log.d(TAG_NAME, "run runLogcat($options)")
                try {
                    val cmd = "logcat $options"
                    val process = Runtime.getRuntime().exec(cmd)
                    val bufferedReader = BufferedReader(InputStreamReader(process.inputStream))
                    var line: String?
                    while (bufferedReader.readLine().also { line = it } != null) {
                        log.append(line)
                    }
                } catch (e: IOException) {
                    log.append("EXCEPTION").append(e)
                }
                result.success(log.toString())
            }

            else -> {
                result.notImplemented()
                return
            }
        }
    }

    private fun closePrevious() {
        if (logcatProcess != null) {
            logcatProcess!!.destroyForcibly()
            logcatProcess = null
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        channel!!.setMethodCallHandler(null)
        eventChannel!!.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, eventSink: EventSink?) {
        this.eventSink = eventSink
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun logcatMonitor(logcatOptions: String) {
        val logcatCmd = "logcat $logcatOptions"
        try {
            Log.d(TAG_NAME, "running command: $logcatCmd")
            logcatProcess = Runtime.getRuntime().exec(logcatCmd)
            val bufferedReader = BufferedReader(InputStreamReader(logcatProcess!!.inputStream))
            var line: String?
            var startTime = SystemClock.elapsedRealtime()
            while (bufferedReader.readLine().also { line = it } != null) {
                val timeInterval = SystemClock.elapsedRealtime() - startTime
                if (timeInterval > sleepIntervalThread) Thread.sleep(sleepTimeThread)
                sendEvent(line)
                startTime = SystemClock.elapsedRealtime()
            }
        } catch (e: IOException) {
            sendEvent("EXCEPTION$e")
        } catch (e: InterruptedException) {
            sendEvent("logcatMonitor interrupted")
        }
        Log.d(TAG_NAME, "closed command: $logcatCmd")
    }

    private fun sendEvent(message: String?) {
        uiThreadHandler.post { eventSink?.success(message) }
    }
}


android/src/main/kotlin/com/josephcrowell/flutter_logcat_monitor/FlutterLogcatMonitorPlugin.kt
package com.josephcrowell.flutter_logcat_monitor

import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.EventSink
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class FlutterLogcatMonitorPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private val uiThreadHandler = Handler(Looper.getMainLooper())
    private var channel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventSink? = null
    private var logcatProcess: Process? = null

    companion object {
        private const val TAG_NAME = "FlutterLogcatMonitorPlugin"
        private const val sleepIntervalThread: Long = 1000
        private const val sleepTimeThread: Long = 200
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_logcat_monitor/methods")
        channel!!.setMethodCallHandler(this)
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_logcat_monitor/events")
        eventChannel!!.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        var options: String?
        val log = StringBuffer()
        options = call.argument("options")
        if (options.isNullOrEmpty()) {
            options = ""
        }
        when (call.method) {
            "startMonitor" -> {
                closePrevious()
                Log.d(TAG_NAME, "run startMonitor()")
                Thread {
                    logcatMonitor(options)
                }.start()
                result.success(true)
            }

            "stopMonitor" -> {
                Log.d(TAG_NAME, "run stopMonitor()")
                closePrevious()
                result.success(true)
            }

            "runLogcat" -> {
                Log.d(TAG_NAME, "run runLogcat($options)")
                try {
                    val cmd = "logcat $options"
                    val process = Runtime.getRuntime().exec(cmd)
                    val bufferedReader = BufferedReader(InputStreamReader(process.inputStream))
                    var line: String?
                    while (bufferedReader.readLine().also { line = it } != null) {
                        log.append(line)
                    }
                } catch (e: IOException) {
                    log.append("EXCEPTION").append(e)
                }
                result.success(log.toString())
            }

            else -> {
                result.notImplemented()
                return
            }
        }
    }

    private fun closePrevious() {
        if (logcatProcess != null) {
            logcatProcess!!.destroyForcibly()
            logcatProcess = null
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        channel!!.setMethodCallHandler(null)
        eventChannel!!.setStreamHandler(null)
    }

    override fun onListen(arguments: Any?, eventSink: EventSink?) {
        this.eventSink = eventSink
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun logcatMonitor(logcatOptions: String) {
        val logcatCmd = "logcat $logcatOptions"
        try {
            Log.d(TAG_NAME, "running command: $logcatCmd")
            logcatProcess = Runtime.getRuntime().exec(logcatCmd)
            val bufferedReader = BufferedReader(InputStreamReader(logcatProcess!!.inputStream))
            var line: String?
            var startTime = SystemClock.elapsedRealtime()
            while (bufferedReader.readLine().also { line = it } != null) {
                val timeInterval = SystemClock.elapsedRealtime() - startTime
                if (timeInterval > sleepIntervalThread) Thread.sleep(sleepTimeThread)
                sendEvent(line)
                startTime = SystemClock.elapsedRealtime()
            }
        } catch (e: IOException) {
            sendEvent("EXCEPTION$e")
        } catch (e: InterruptedException) {
            sendEvent("logcatMonitor interrupted")
        }
        Log.d(TAG_NAME, "closed command: $logcatCmd")
    }

    private fun sendEvent(message: String?) {
        uiThreadHandler.post { eventSink?.success(message) }
    }
}
